client-core/src/service/wallet_service.rs

```diff
@@ -810,12 +855,44 @@ where
         let info_keyspace = get_info_keyspace(name);
         let roothash_count: u64 =
             read_number(&self.storage, &info_keyspace, "roothashindex", None)?;
-        for i in 0..roothash_count {
-            let value = self.storage.get(&roothash_keyspace, format!("{}", i))?;
-            if let Some(raw_value) = value {
-                let mut roothash_found: H256 = H256::default();
-                roothash_found.copy_from_slice(&raw_value);
-                ret.insert(roothash_found);
+
+        if reversed {
+            let mut start = 0;
+            if roothash_count > offset {
+                start = roothash_count - 1 - offset;
+            }
+            let mut end = 0;
+            if limit > 0 && start >= (limit - 1) {
+                end = start - (limit - 1);
+            }
+            let mut i = start;
+            while i >= end {
+                let value = self.storage.get(&roothash_keyspace, format!("{}", i))?;
+                if let Some(raw_value) = value {
+                    let mut roothash_found: H256 = H256::default();
+                    roothash_found.copy_from_slice(&raw_value);
+                    ret.insert(roothash_found);
+                }
+                if 0 == i {
+                    break;
+                }
+
+                // iterate
+                i -= 1;
+            }
+        } else {
+            let start = offset;
+            let mut end = roothash_count;
+            if limit > 0 {
+                end = std::cmp::min(offset + limit, roothash_count);
+            }
+            for i in start..end {
+                let value = self.storage.get(&roothash_keyspace, format!("{}", i))?;
+                if let Some(raw_value) = value {
+                    let mut roothash_found: H256 = H256::default();
+                    roothash_found.copy_from_slice(&raw_value);
+                    ret.insert(roothash_found);
+                }
             }
         }
```

Comment:
seems this code is exactly the same as with staking_keys (the only difference is the key space) -- perhaps can be extracted out to a helper function used by both?

Reply:
ok